pkg/connector/ble/ble.go

```diff
@@ -167,56 +201,98 @@ func ScanVehicleBeacon(ctx context.Context, vin string) (Advertisement, error) {
 	return a, nil
 }
 
-func scanVehicleBeacon(ctx context.Context, localName string) (Advertisement, error) {
-	var err error
-	ctx2, cancel := context.WithCancel(ctx)
-	defer cancel()
-
-	ch := make(chan Advertisement, 1)
-	fn := func(a Advertisement) {
-		if a.LocalName() != localName {
-			return
-		}
-		select {
-		case ch <- a:
-			cancel() // Notify device.Scan() that we found a match
-		case <-ctx2.Done():
-			// Another goroutine already found a matching advertisement. We need to return so that
-			// the MacOS implementation of device.Scan(...) unblocks.
+func scanVehicleBeacon(ctx context.Context, localName string) (*ScanResult, error) {
+	scanIsStopped := false
+	stopScan := func() {
+		scanIsStopped = true
+		if err := adapter.StopScan(); err != nil {
+			log.Warning("ble: failed to stop scan: %s", err)
 		}
 	}
 
-	if err = device.Scan(ctx2, false, fn); !errors.Is(err, context.Canceled) {
-		// If ctx rather than ctx2 was canceled, we'll pick that error up below. This is a bit
-		// hacky, but unfortunately device.Scan() _always_ returns an error on MacOS because it does
-		// not terminate until the provided context is canceled.
-		return nil, err
+	errorCh := make(chan error)
+	foundCh := make(chan *ScanResult)
+
+	// FIXME: We need to check if the context is already done to prevent starting the scan
+	// and then immediately stopping it, since that would create a chance for
+	// the scan to not be stopped before being started. There is still a small chance that
+	// the scan will not be stopped properly if the context is done between
+	// the check and the start of the scan, but that is a very small window.
+	// This can be fixed once the bluetooth library supports context handling.
+	// See: https://github.com/tinygo-org/bluetooth/issues/339
+	if ctx.Err() != nil {
+		return nil, ctx.Err()
 	}
 
-	select {
-	case a, ok := <-ch:
-		if !ok {
-			// This should never happen, but just in case
-			return nil, fmt.Errorf("scan channel closed")
+	// Another fix: We need to wait for the Scan() call to finish before returning because
+	// waiting for StopScan() to finish is not enough.
+	scanFinished := make(chan struct{})
+	defer func() {
+		<-scanFinished
+	}()
+
+	// Scan is blocking, so run it in a goroutine
+	go func() {
+		defer close(scanFinished)
+		if scanIsStopped {
+			return
 		}
-		return a, nil
+		log.Debug("Scanning for %s...", localName)
+		if err := adapter.Scan(func(_ *bluetooth.Adapter, result bluetooth.ScanResult) {
+			// If we have stopped the scan and we still get results it means
+			// that the case described in the comment above has happened.
+			if scanIsStopped {
+				stopScan()
+				return
+			}
+
+			if result.LocalName() == localName {
```

Comment:
This code has the same issues described in https://github.com/teslamotors/vehicle-command/pull/373#discussion_r1956722033.